### PR TITLE
Implement Grape::API.recognize_path

### PR DESCRIPTION
**Pull request by namusyaka**
ref #1072 
I've tried to implement `Grape::DSL::InsideRoute#recognize_path`.
It returns value conforming the format like `[original pattern, expanded params]`.

This is a first step. I'd like to hear your opinions.
Thanks.

### CODE REVIEW

File: lib/grape/endpoint.rb [hunk 1]
```diff
@@ -120,6 +120,7 @@ def reset_routes!
     end
 
     def mount_in(router)
+      @router ||= router
```
Comment (dblock) [inline lib/grape/endpoint.rb]: We didn't address mounting an endpoint in multiple places, but eventually we will. So maybe we don't want `@router`?

File: spec/grape/dsl/inside_route_spec.rb [hunk 2]
```diff
@@ -357,6 +357,33 @@ def initialize
...
+          env = Rack::MockRequest.env_for('/')
+          expect(subject.call(env)[2].body[0]).to eq(['/test/:id', { id: '1234', format: 'json' }].to_json)
+        end
+
+        describe 'returned params' do
+          it 'is able to access params by using string or symbol as a key' do
+            subject.get('/test/:id'){}
+            subject.get do
+              _, params = recognize_path('/test/1234')
+              "string: #{params['id']}, symbol: #{params[:id]}"
```
Comment (dblock) [inline spec/grape/dsl/inside_route_spec.rb]: Just for the sake of sanity in the test, return json.

**Comment by dblock** [comment]
It's on the right track, but shouldn't `recognize_path` return an endpoint?

File: lib/grape/endpoint.rb (hunk 1 above)
Comment (namusyaka) [inline lib/grape/endpoint.rb]: Yeah, right. Removed.

**Comment by namusyaka** [comment]
@dblock Now `recognize_path` returns an instance of `Grape::Endpoint`. How about this?

**Comment by dblock** [comment]
Very good. I think the last thing we need is a README entry that explains how you can use `recognize_path` from the outside of an API, so given a `API < Grape::API`, calling `recognize_path` to get a endpoint that would match.

**Comment by namusyaka** [comment]
>> I think the last thing we need is a README entry that explains how you can use recognize_path from the outside of an API, so given a API < Grape::API, calling recognize_path to get a endpoint that would match.

Hmm, makes sense. If I understand thing correctly, `recognize_path` should be implemented as a class method of `Grape::API`.
Maybe we can't call `recognize_path` outside of an API, without tricky code.
So I moved the method place from `InsideRoute` to `API`.
Could you please check this?

**Comment by dblock** [comment]
Perfect, merging.

**Pull request merged**